frame/support/src/migrations.rs

```diff
@@ -15,19 +15,126 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#[cfg(feature = "try-runtime")]
-use crate::storage::unhashed::contains_prefixed_key;
 use crate::{
-	traits::{GetStorageVersion, NoStorageVersionSet, PalletInfoAccess, StorageVersion},
+	traits::{
+		GetStorageVersion, NoStorageVersionSet, OnRuntimeUpgrade, PalletInfoAccess, StorageVersion,
+	},
 	weights::{RuntimeDbWeight, Weight},
 };
 use impl_trait_for_tuples::impl_for_tuples;
 use sp_core::Get;
 use sp_io::{hashing::twox_128, storage::clear_prefix, KillStorageResult};
 use sp_std::marker::PhantomData;
+
 #[cfg(feature = "try-runtime")]
 use sp_std::vec::Vec;
 
+/// Make it easier to write versioned runtime upgrades.
+///
+/// VersionedRuntimeUpgrade allows developers to write migrations without worrying about checking
+/// and setting storage versions. Instead, the developer wraps their migration in this struct which
+/// takes care of version handling using best practices.
+///
+/// It takes 4 type parameters:
+/// - `Version`: The version being upgraded *from*.
+/// - `Inner`: An implementation of `OnRuntimeUpgrade`.
+/// - `Pallet`: The Pallet being upgraded.
+/// - `Weight`: The runtime's RuntimeDbWeight implementation.
+///
+/// `Version` is used to check if the migration should be run. If it is run, the pallets on-chain
+/// version is incremented.
+///
+/// Example:
+/// ```ignore
+/// parameter_types! {
+/// 	pub const V4ToV5: StorageVersion = StorageVersion::new(4);
+/// 	pub const V5ToV6: StorageVersion = StorageVersion::new(5);
+/// }
+///
+/// // Migrations to pass to the Executive pallet.
+/// pub type Migrations = (
+/// 	// ...other migrations
+/// 	VersionedRuntimeUpgrade<
+/// 		V4ToV5,
+/// 		parachains_configuration::migration::v5::MigrateToV5<Runtime>,
+/// 		Configuration,
+/// 		RocksDbWeight,
+/// 	>,
+/// 	VersionedRuntimeUpgrade<
+/// 		V5ToV6,
+/// 		parachains_configuration::migration::v6::MigrateToV6<Runtime>,
+/// 		Configuration,
+/// 		RocksDbWeight,
+/// 	>,
+/// 	// ...other migrations
+/// );
+/// ```
+pub struct VersionedRuntimeUpgrade<Version, Inner, Pallet, Weight> {
+	_marker: PhantomData<(Version, Inner, Pallet, Weight)>,
+}
+
+/// Implementation of the `OnRuntimeUpgrade` trait for `VersionedRuntimeUpgrade`.
+///
+/// Primarily, it
+/// 1. Logs a warning in `pre_upgrade` if the storage version specified in the pallet appears to be
+/// misconfigured before returning `Inner::pre_upgrade`.
+///
+/// 2. Performs the actual runtime upgrade in `on_runtime_upgrade` only if the on-chain version of
+/// the pallets storage matches `Version`. If the versions do not match, it writes a log notifying
+/// the developer that the migration is a noop.
+impl<
+		Version: Get<StorageVersion>,
```

Comment:
Did you try out how it looks like with an associated constant?  
I think the `Get` always makes for an awkward usage :seedling: 

How would this be used btw?  
type MigrateToV3 = VersionedRuntimeUpgrade<ConstU32<3>, InnerMigrationToV3, Pallet, Weights>`?  
It is slightly odd that it is implemented for all `Version`, but not a problem.